pkg/testutil/componentsuite/validate.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/converter"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/options/patchtmpl"
+	"github.com/GoogleCloudPlatform/k8s-cluster-bundle/pkg/validate"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type objKey struct {
+	name string
+	kind string
+}
+
+// String returns the stringified objKey.
+func (o objKey) String() string {
+	return fmt.Sprintf("{name: %s, kind: %s}", o.name, o.kind)
+}
+
+func objKeyFromObjCheck(oc ObjectCheck) objKey {
+	return objKey{
+		name: oc.Name,
+		kind: oc.Kind,
+	}
+}
+
+func objKeyFromObj(oc *unstructured.Unstructured) objKey {
+	return objKey{
+		name: oc.GetName(),
+		kind: oc.GetKind(),
+	}
+}
+
+func runValidate(t *testing.T, comp *bundle.Component, tc *TestCase) {
+	if errList := validate.Component(comp); len(errList) > 0 {
+		t.Errorf("There were errors validating component:\n%v", errList.ToAggregate())
+	}
+
+	if tc.Expect.CanKubeDeserialize {
+		checkObjectsDeserialize(t, comp)
+	}
+
+	checkObjectProperties(t, comp, tc)
+}
+
+func checkObjectsDeserialize(t *testing.T, comp *bundle.Component) {
+	scheme := patchtmpl.DefaultPatcherScheme()
+	deserializer := scheme.Codecs.UniversalDeserializer()
+
+	for _, obj := range comp.Spec.Objects {
+		key := objKeyFromObj(obj)
+		objByt, err := converter.FromObject(obj).ToJSON()
+		if err != nil {
+			// This would be pretty unlikely
+			t.Errorf("Error converting object %v to JSON: %v", key, err)
+			continue
+		}
+
+		_, err = runtime.Decode(deserializer, objByt)
+		if err != nil {
+			t.Errorf("Error decode object %v: %v", key, err)
+		}
+	}
+}
+
+func checkObjectProperties(t *testing.T, comp *bundle.Component, tc *TestCase) {
+	objCheckMap := make(map[objKey]ObjectCheck)
+	for _, oc := range tc.Expect.Objects {
+		objCheckMap[objKeyFromObjCheck(oc)] = oc
+	}
+
+	objMap := make(map[objKey]string)
+	for _, obj := range comp.Spec.Objects {
+		matchFail := false
+
+		key := objKeyFromObj(obj)
+		objStr, err := converter.FromObject(obj).ToYAMLString()
+		if err != nil {
+			// This is a very unlikely error.
+			t.Fatal(err)
+		}
+		objMap[key] = objStr
+
+		check := objCheckMap[key]
+		for _, expStr := range check.FindSubstrs {
+			if !strings.Contains(objStr, expStr) {
+				t.Errorf("Did not find %q in object %v, but expected to", expStr, key)
+				matchFail = true
+			}
+		}
+
+		for _, noExpStr := range check.NotFindSubstrs {
+			if strings.Contains(objStr, noExpStr) {
+				t.Errorf("Found %q in object %v, but did not expect to", noExpStr, key)
+				matchFail = true
+			}
+		}
+
+		if matchFail {
+			t.Logf("Contents for object that didn't meet expectations %v:\n%s", key, objStr)
+		}
+	}
+
+	for key := range objCheckMap {
+		if _, ok := objMap[key]; !ok {
+			t.Errorf("Got object-keys %q, but expected to find object %v", stringMapKeys(objMap), key)
+		}
+	}
+}
+
+func stringMapKeys(m map[objKey]string) []string {
```

Comment:
You can use sets.StringKeySet.

Reply:
/shrug. I think it's a little heavy weight to rely on K8S ApiMachinery here personally.